msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AcquireTokenSilentSupplier.java

```diff
@@ -120,4 +82,74 @@ AuthenticationResult execute() throws Exception {
+                RefreshTokenParameters.builder(silentRequest.parameters().scopes(), cachedResult.refreshToken()).build(),
+                silentRequest.application(),
+                silentRequest.requestContext(),
+                silentRequest);
+
+        AcquireTokenByAuthorizationGrantSupplier acquireTokenByAuthorisationGrantSupplier =
+                new AcquireTokenByAuthorizationGrantSupplier(clientApplication, refreshTokenRequest, requestAuthority);
+
+        try {
+            AuthenticationResult refreshedResult = acquireTokenByAuthorisationGrantSupplier.execute();
+
+            refreshedResult.metadata().tokenSource(TokenSource.IDENTITY_PROVIDER);
+
+            log.info("Access token refreshed successfully.");
+            return refreshedResult;
+        } catch (MsalServiceException ex) {
+            //If the token refresh attempt threw a MsalServiceException but the refresh attempt was done
+            // only because of refreshOn, then simply return the existing cached token
+            if (!afterRefreshOn && (silentRequest.parameters().forceRefresh() || StringHelper.isBlank(cachedResult.accessToken()))) {
+                throw ex;
+            }
+            return cachedResult;
+        }
+    }
+
+    private void setRefreshTelemetry(AuthenticationResult cachedResult) {
+        //As of version 3 of the telemetry schema, there is a field for collecting data about why a token was refreshed,
+        // so here we set the telemetry value based on the cause of the refresh
+        if (silentRequest.parameters().forceRefresh()) {
+            clientApplication.serviceBundle().getServerSideTelemetry().getCurrentRequest().cacheInfo(
+                    CacheTelemetry.REFRESH_FORCE_REFRESH.telemetryValue);
+        } else if (afterRefreshOn) {
+            clientApplication.serviceBundle().getServerSideTelemetry().getCurrentRequest().cacheInfo(
+                    CacheTelemetry.REFRESH_REFRESH_IN.telemetryValue);
+        } else if (cachedResult.expiresOn() < new Date().getTime() / 1000) {
+            clientApplication.serviceBundle().getServerSideTelemetry().getCurrentRequest().cacheInfo(
+                    CacheTelemetry.REFRESH_ACCESS_TOKEN_EXPIRED.telemetryValue);
+        } else if (StringHelper.isBlank(cachedResult.accessToken())) {
+            clientApplication.serviceBundle().getServerSideTelemetry().getCurrentRequest().cacheInfo(
+                    CacheTelemetry.REFRESH_NO_ACCESS_TOKEN.telemetryValue);
+        }
+    }
+
+    //Handles any logic to determine if a token should be refreshed, based on the request parameters and the status of cached tokens
+    private boolean shouldRefresh(SilentParameters parameters, AuthenticationResult cachedResult) {
+
+        //If there is a refresh token but no access token, we should use the refresh token to get the access token
+        if (StringHelper.isBlank(cachedResult.accessToken()) && !StringHelper.isBlank(cachedResult.refreshToken())) {
```

Comment:
highly unlikely, I recommend moving to the bottom as optimization.

Reply:
Yeah, the only reasonable scenario I can see this one happening is if someone configures a cache where they wiped all of the access tokens but left the refresh tokens.

In the latest commit I moved the check for forceRefresh to the top (since it can simply override any other option if true) and this check to the bottom.